### ISSUE TITLE
[Proposal] : Self-validating models

### ISSUE THREAD

**Issue by reshadman** [laravel/ideas#1167]
Laravel has a great validation library, It would be nice to allow developers to use **self-validating** models, there are multiple benefits for this approach:
 - It will be lead to less code, as Laravel encourages resource based design, most of the time the model is identical to the user input, also developer may define different behaviour for different contexts (such as `admin`, `update` and ...).
 - The current validation progress seems a little bit procedural (First I validate,then I perform some operations), this could be moved to a more declarative way, which also encourages resource based design, and DRY. 
- It also helps to keep the needed logic into the models themselves, instead of applying complex and none productive approaches,  Like defining standalone `UserValidator` inside a service object, **most** of the time you are interacting with your resource and there is not any logic outside the model and its friends (other related models), the problem encounters when we tend to wrap our procedural mess inside a service object which is controlled by an standalone validator, but if we would have designed the models correctly, there wouldn't be any external logic, which then the external homeless procedure is wrapped in to a service object , or a repositoy class (which I don't know why it is called a repositoy class), I have been in a project that every resource action has been wrapped inside a service object, it is still a mess, and the excuse that they are reused is correct for **%5** of the cases, and in those case you still may need to write a switch case in the service object to behave on two different inputs, This self-validating features among with other infrastructure level features that Laravel offers such as notifications, can lead you to write the logic into the models, no standalone service objects, no repositories, no standalone validators. 

I have prepared a simple interface for applying this feature to the model
```php
<?php


class Task extends Model
{
     protected static $rules = [
        'title' => 'required|string',
        'content' => 'required|string|current:not_too_many_mentions', // It is possible to call model methods to add behaviour to the validation.
        'assignee' => 'required|current:assignee_is_admin', // This is a needed relationship for the model, it can be more explicit with a rule like `related`
        'creator' => 'required|delegate|' // This will also delegate the validation of the realtion if needed to its associated model.
    ];

    protected static $adminRules = [
        //
    ];

    public function validateNotTooManyMentions()
    {
        // Count number of mentions in a task
        return true || false;
    }

    public function validateAssigneeIsAdmin()
    {
        return !!optional($this->assignee)->isAdmin();
    }

    public function assignee()
    {
        return $this->belongsTo(User::class, 'assignee_id');
    }

    public function creator()
    {
        return $this->belongsTo(User::class, 'creator_id');
    }
}

// Creating tasks with validating them.
$note = new Note;
$note->fill($request->only($note->getFillable()));
$note->assignee()->associate(auth()->user());
$note->save();

// If there are multiple validation strategies there would be possible
// to define contexts
$note->onContext('admin')->save();
// Or
$note->save(['context' => 'admin']); 

// The note save would return false if the validation fails, we can also
// allow the model to be more strict and throw validation errors. ofcourse
// some needed methods like `fails` , `valid`, `getErrors` should be added 
// to Eloquent.
```

The final implementation can be a standalone trait like `InteractsWithValidator` or `SelfValidating` or `SelfValidates`.

**Comment by Kyslik** [comment]
What about FormRequest? IMO: FormRequest is the best place to validate data from client.

**Comment by sisve** [comment]
There's a proper term for this, but it slips my mind at the moment.

I read this idea as a way to ensure that the model is always in a valid state. You say that it will always have a title, and it will be string. That must be true no matter what happens. This is different from form requests that only validates a specific form submit.

I believe these rules would catch cases where a developer does `$model->name = $request->input('name');` or `$this->model = 1337`. There doesn't have to be a form request, or even a http post, involved at all.

I guess one thing need to be decided; should the model always be in a valid state, or should it allow invalid states temporarily? (Can you do `$model->name = null; $model->name = 'test';`, or should that trigger a state-related exception when setting it to null?)

**Comment by reshadman** [comment]
@Kyslik Form requests are not shared through different contexts. you will end up reading model(resource) behaviour in different places such as form requests, controllers, and wherever that model is somehow interacting with your application but does not use that specific form request. ofcourse it is not a bad practice to validate any data you want, but there is a problem that in a re

**Comment by 36864** [comment]
I thought this sounded familiar: #1083

I like the overall idea, but I think there should be an easy way to temporarily disable validation on a model. Maybe through setting a null context, or more explicitly, `onContext('no-validation')`?

**Comment by reshadman** [comment]
@Kyslik Form requests are great, the bad about them is  that they can not be shared through contexts (such as in backoffice and website). Another problem is that you end up reading model behaviour in different places, and also they carry their name, they tend to validate forms not resources, self validating models can perform other interactions that may help you a lot, such as delegating the validation of a related model, calling validation methods directly in the model, and the luxury of knowing how exactly your controller resource, your command line app, or your cron schedule is mutating the model in one place. The validation can be ditched by an approach like `unguard`, `reguard`.

**Comment by reshadman** [comment]
@sisve Your point was a discussion in our office, I prefer the model to delay the validation until it is told to validate, there is another problem with the entire approach in migrations, in a migration process old records may get to an invalid state, anyway when my resource is identical to my user's input I prefer not to write any extra layer of validating, there are rare cases when the user input can not be handled with this approach using contexts. I like the idea of keeping the model in valid state, in most cases this possible with self validating models. In some cases (Such as a multi step signup without using session) this may be a little complex.

**Comment by reshadman** [comment]
@36864 Yeah, they can be handled using contexts, or like the way a model is being unguarded and reguarded. I think that this approach as a choice with little extra addition to the model, will end up to have more meaningful models for me, I have been using this approach with Rails and I have been happy with that.

**Comment by nCrazed** [comment]
> there is another problem with the entire approach in migrations, in a migration process old records may get to an invalid state

This is a self made problem. The process of data migrations involves going from old version of valid state to a new version of valid state. Since a model will be able to represent only one of those versions (usually the new), it has no place in migration code. The migration code should be dealing with underlying storage directly or at least through abstractions that are below the model (e.g. query builder).

> I like the idea of keeping the model in valid state, in most cases this possible with self validating models. In some cases (Such as a multi step signup without using session) this may be a little complex.

If you have a case for an invalid state than either your model is not correct for your use case, or you need new models. Mult-step forms are a good example for when you'd have additional models to capture the valid intermediary states that are later assembled into a final, complete model.

**Comment by reshadman** [comment]
@nCrazed You are right, if you can manage to achieve the goals you mentioned, It is much cleaner and better.

**Comment by reshadman** [comment]
@donnysim You can handle different conditions with contexts, look at the snippet above. also it is much simpler to handle validation models that are interacted with models instead of explicitly saying `unique:users,email,12` this is handled automaticaly by the model. 

You can also include model behaviour in to the validation rules, like the above one I mentioned (The assignee of the task must be an admin for example).

**Comment by salomoni** [comment]
I like the idea and the contextual validation would be a very nice touch. It still somehow feels gross to add the validation rules and methods to the model.

Does it validate the model only on save? In the example the `$note` is invalid in begin as it doesn't have any values so throwing exceptions there is a no-no.

There would still be place for the form requests e.g. validating unique emails.

**Comment by vikas5914** [comment]
It's like Typechecking With PropTypes for Laravel.

**Comment by reshadman** [comment]
@salomoni I think there is room for validation rules inside models, in my opinion they increase model readablity and self description as long as models are simple resources of our application it would be nice to keep them always in a valid state and know what the valid state is, however in most use cases they are identical to form definitions and there would be no need to add extra validation to check input, but there is always ability to define input validators and temporary disable model validations in case we need one.

Ofcourse there would be some sugar when validating inside a model, some checks like unique email rule except the current model can be automatically  handled by the model without excluding the id explicitly. You can use if/unless returning  model methods as validation rules with messages related to the check right inside where the domain actually is, you can explicitely define which rule is excuted on which context via defining if/unless methods for the model. 


Relationship rules can be defined, too, to check their existence or even delegating the job to self validate them and aggregate the validation errors to the deligator to have an Error bag tree, 
Rails has been using this method and everyone is happy with this approach.

Finally I think such an approach will lead to less code and more declarative and rich models, with things like Service objects, repositories and dedicated form requests or validator objects, we are making anemic models, models with just database data mapping definitions and setter getters, most of the work like validation can be delegated to models themeselves.
